### PR TITLE
Above ground zoi

### PR DESCRIPTION
**Pull request by zaphod-bx**
sooooooooooo,
this may work! check it out now!

### CODE REVIEW

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/__init__.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Thu Nov  8 15:25:03 2018
+
+@author: bathmann
```
Comment (jbathmann) [inline TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/__init__.py]: Truly? ;)

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+@date: 2018-Today
+@author: jasper.bathmann@ufz.de
```
Comment (jbathmann) [inline TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: author...

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                "Tag(s) " + string +
+                "are not given for above-ground grid initialisation in project file.")
+        l_x = x_2 - x_1
+        l_y = y_2 - y_1
+        x_step = l_x / x_resolution
+        y_step = l_y / y_resolution
+        self.min_r_crown = np.mean([x_step, y_step]) * 2**0.5 
+        xe = np.linspace(x_1+x_step/2.,x_2-x_step/2., x_resolution ,endpoint=True) 
+        ye = np.linspace(y_1+y_step/2.,y_2-y_step/2., y_resolution ,endpoint=True) 
+        self.my_grid = np.meshgrid(xe,ye)
```
Comment (jbathmann) [inline TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: nice

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        for i in range(len(self.xe)):
+            self.tree_win.append(sum(sum(self.winner==i)))
+        self.aboveground_resources = np.divide(self.tree_win,self.crown_area)
+        
+    ## This function calculates the tree height at a (mesh-)point depending
+    #  on the distance from the tree position.\n
+    #  @param hst - stem height\n
+    #  @param rcr - crown radius\n
+    #  @param distance - distance from the stem position
+    def calculateHeightFromDistance(self,hst,rcr,distance):
```
Comment (jbathmann) [inline TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: the parameter names are not instructive.
please change hst -> h_stem ....

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        height = rcr-distance
+        height = np.maximum(height, 0,height)
```
Comment (jbathmann) [inline TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: what do you do here?

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        height[height>0] = hst + (4*rcr**2 - distance[height>0]**2)**0.5
+        return height
+
+    ## This function reads x- and y-domain and mesh resolution
+    #  from project file and creates the mesh.\n
+    def makeGrid(self,args):
```
Comment (jbathmann) [inline TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: please document argument

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        self.t_ini = t_ini
+        self.t_end = t_end
+        self.canopy_height = np.zeros_like(self.my_grid[0])
+        self.winner = np.full_like(self.my_grid[0],fill_value=np.nan)
+
+    ## Before being able to calculate the resources, all tree entities need
+    #  to be added with their current implementation for the next timestep.
+    #  @param position, geometry, parameter
+    def addTree(self, x, y, geometry, parameter):
+        self.trees.append(1)
```
Comment (jbathmann) [inline TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: do you need this here?

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+    ## This functions prepares arrays for the competition
+    #  concept. In the SimpleAssymmetricZOI concept, trees geometric measures
+    #  are saved in simple lists and the timestepping is updated. Two numpy
+    #  arrays similar to the mesh array for canopy height and tree ID of highest tree
+    #  at corresponding mesh point are initialised. \n
+    #  @param t_ini - initial time for next timestep \n
+    #  @param t_end - end time for next timestep
+    def prepareNextTimeStep(self, t_ini, t_end):
+        self.trees = []
```
Comment (jbathmann) [inline TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: I think you dont need this?

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    def __init__(self, args):
+        case = args.find("type").text
+        print("Initiate aboveground competition of type " + case + ".")
+        self.makeGrid(args)
+
+    ## This function returns the AbovegroundResources calculated in the
+    #  subsequent timestep.\n
+    #  @return: np.array with $N_tree$ scalars
+    def calculateAbovegroundResources(self):
+        for i in range(len(self.xe)):
```
Comment (jbathmann) [inline TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: canopy_heigth is only filled with zeros so far. guess this must be added to add_tree

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/__init__.py (hunk 1 above)
Comment (zaphod-bx) [reply TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/__init__.py]: well, okay, I'll authorise me!

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py (hunk 2 above)
Comment (zaphod-bx) [reply TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: dito

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py (hunk 4 above)
Comment (zaphod-bx) [reply TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: h_stem gibts schon, bzw, self.h_stem
daher: stem_height
besser?

Comment (zaphod-bx) [reply TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: und rcr dito

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py (hunk 5 above)
Comment (zaphod-bx) [reply TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: What do you mean?
    ## This function calculates the tree height at a (mesh-)point depending
    #  on the distance from the tree position.\n
or do you mean  lines 42/43 in particular?
that is getting 0 if the meshpoint isn't under the tree's crown and something > 0 if it is.

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py (hunk 6 above)
Comment (zaphod-bx) [reply TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: Well, okay.

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py (hunk 7 above)
Comment (zaphod-bx) [reply TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: Nope

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py (hunk 8 above)
Comment (zaphod-bx) [reply TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: Me too!

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py (hunk 9 above)
Comment (zaphod-bx) [reply TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: No, it is not the crown height but the height of the stand canopy at all mesh points. Thus, it is defined in prepareNextTimeStep

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py (hunk 5 above)
Comment (jbathmann) [reply TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: but why are you taking the maximum of three values?

File: TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    ## This function calculates the tree height at a (mesh-)point depending
+    #  on the distance from the tree position.\n
+    #  @param hst - stem height\n
+    #  @param rcr - crown radius\n
+    #  @param distance - distance from the stem position
+    def calculateHeightFromDistance(self,hst,rcr,distance):
+        height = rcr-distance
+        height = np.maximum(height, 0,height)
+        height[height>0] = hst + (4*rcr**2 - distance[height>0]**2)**0.5
+        return height
```
Comment (jbathmann) [inline TreeModelLib/AbovegroundCompetition/SimpleAsymmetricZOI/SimpleAsymmetricZOI.py]: Hey, I just encountered that this loop might become very slow as soon as the domain grows.
Here is a suggestion how to increase speed. However, this approach is also limited. On my laptop a setup consisting of 750x750 grid points and 400 trees is the maximum, which can be allocated (but my memory is quite small...). However, allocation one timestep need approx 4s, which I think is pretty good for the setup size.
WDYT?

`




















    def calculateAbovegroundResources(self):
        #for i in range(len(self.xe)):
        t_start = time.time()
        distance = (((self.my_grid[0][:,:,np.newaxis] - np.array(self.xe)[np.newaxis,np.newaxis,:])**2
                     + (self.my_grid[1][:,:,np.newaxis]-np.array(self.ye)[np.newaxis,np.newaxis,:])**2)**0.5)
        t_dist = time.time()
        my_height = self.calculateHeightFromDistance(np.array(self.h_stem),np.array(self.r_crown),distance)
        t_height = time.time()
        #Define empty array of dimensions res_x, res_y, n_trees
        crown_areas = np.zeros_like(my_height)
        #Add a one, where tree is larger than 0
        crown_areas[np.where(my_height>0)] +=1
        #Count all nodes, which are occupied by trees
        #returns array of shape (ntrees)
        crown_areas = crown_areas.sum(axis=(0,1))

        # look for largest tree
        self.canopy_height = np.max(my_height, axis=-1)
        self.canopy_height[np.where(np.less(self.canopy_height,0))] = 0
        #print(np.shape(my_height.sum(axis=-1)))
        #print(np.shape(my_height.sum(axis=0)))
        #crown_area.append(sum(sum(my_height > 0)))

        #define array to count wins
        wins = np.zeros_like(my_height)
        #indicate, where tree is highest
        wins[np.where(np.equal(my_height,self.canopy_height[:,:,np.newaxis]))] +=1

        #Account for shared wins
        cumwins = wins.sum(axis=-1)
        cumwins[np.where(cumwins==0)] = -1
        wins = wins/ cumwins[:,:,np.newaxis]
        wins[np.where(wins<0)] = 0
        #Count number of wins
        wins=wins.sum(axis=(0,1))
        t_end = time.time()
        self.aboveground_resources = wins/crown_areas
        print("t_dist:", (t_dist - t_start)," in %", (t_dist - t_start)/(t_end-t_start))
        print("t_h_alloc:", (t_height - t_dist)," in %", (t_height - t_dist)/(t_end-t_start))
        print("t_dist:", (t_end - t_height)," in %", (t_end - t_height)/(t_end-t_start))


    ## This function calculates the tree height at a (mesh-)point depending
    #  on the distance from the tree position.\n
    #  @param hst - stem height\n
    #  @param rcr - crown radius\n
    #  @param distance - distance from the stem position
    def calculateHeightFromDistance(self,hst,rcr,distance):
        bools = rcr[np.newaxis, np.newaxis,:]>distance
        idx = np.where(bools)
        height = np.full_like(distance, fill_value = -99999)
        height[idx] = hst[idx[2]] + (4*rcr[idx[2]]**2 - distance[idx]**2)**0.5
        return height`

**Pull request merged**